Firestore/core/src/firebase/firestore/local/leveldb_key.cc

```diff
@@ -0,0 +1,764 @@
+    if (label == ComponentLabel::Terminator) {
+      is_terminated = true;
+      contents = tmp;
+      break;
+    }
+
+    // Reset tmp since all the different read routines expect to see the
+    // separator first
+    tmp = contents;
+
+    if (label == ComponentLabel::PathSegment) {
+      DocumentKey document_key;
+      if (!ReadDocumentKey(&tmp, &document_key)) {
+        break;
+      }
+      absl::StrAppend(&description,
+                      " key=", document_key.path().CanonicalString());
+
+    } else if (label == ComponentLabel::TableName) {
+      std::string table;
+      if (!ReadLabeledString(&tmp, ComponentLabel::TableName, &table)) {
+        break;
+      }
+      absl::StrAppend(&description, table, ":");
+
+    } else if (label == ComponentLabel::BatchId) {
+      model::BatchId batch_id;
+      if (!ReadBatchID(&tmp, &batch_id)) {
+        break;
+      }
+      absl::StrAppend(&description, " batch_id=", batch_id);
+
+    } else if (label == ComponentLabel::CanonicalId) {
+      std::string canonical_id;
+      if (!ReadCanonicalID(&tmp, &canonical_id)) {
+        break;
+      }
+      absl::StrAppend(&description, " canonical_id=", canonical_id);
+
+    } else if (label == ComponentLabel::TargetId) {
+      model::TargetId target_id;
+      if (!ReadTargetID(&tmp, &target_id)) {
+        break;
+      }
+      absl::StrAppend(&description, " target_id=", target_id);
+
+    } else if (label == ComponentLabel::UserId) {
+      std::string user_id;
+      if (!ReadUserID(&tmp, &user_id)) {
+        break;
+      }
+      absl::StrAppend(&description, " user_id=", user_id);
+
+    } else {
+      absl::StrAppend(&description, " unknown label=", static_cast<int>(label));
+      break;
+    }
+
+    contents = tmp;
+  }
+
+  if (contents.size() > 0) {
+    absl::StrAppend(&description, " invalid key=<", InvalidKey(key), ">");
+
+  } else if (!is_terminated) {
+    absl::StrAppend(&description, " incomplete key");
+  }
+
+  absl::StrAppend(&description, "]");
+  return description;
+}
+
+std::string LevelDbVersionKey::Key() {
+  std::string result;
+  WriteTableName(&result, kVersionGlobalTable);
+  WriteTerminator(&result);
+  return result;
+}
+
+std::string LevelDbMutationKey::KeyPrefix() {
+  std::string result;
+  WriteTableName(&result, kMutationsTable);
+  return result;
+}
+
+std::string LevelDbMutationKey::KeyPrefix(absl::string_view user_id) {
+  std::string result;
+  WriteTableName(&result, kMutationsTable);
+  WriteUserID(&result, user_id);
+  return result;
+}
+
+std::string LevelDbMutationKey::Key(absl::string_view user_id,
+                                    model::BatchId batch_id) {
+  std::string result;
+  WriteTableName(&result, kMutationsTable);
+  WriteUserID(&result, user_id);
+  WriteBatchID(&result, batch_id);
+  WriteTerminator(&result);
+  return result;
+}
+
+bool LevelDbMutationKey::Decode(leveldb::Slice key) {
+  user_id_.clear();
+
+  return ReadTableNameMatching(&key, kMutationsTable) &&
```

Comment:
I wish some of this boilerplate could be reduced, but don't really have a good suggestion. The best I can come up with is having a helper method that represents each of the Key classes as a tuple, and then Encode/Decode methods could delegate to a variadic template function like `void LevelDbMutationKey::Encode() { EncodeImpl(AsTuple(*this)); }` However, it would require too much reworking, and will also increase binary size somewhat.

Reply:
The gain would be removing the duplication between `Decode` and `Key` (the common source for errors in serializing/deserializing is when those get out of sync) - both would delegate the tuple representation to a common helper. But yeah, it still won't outweigh the downsides.

Reply:
I see what you mean. That's a significant chunk of work and I'm still not entirely sure how I'd make partial keys work there.